### ISSUE TITLE
how to get port speed

### ISSUE THREAD

**Issue by hubertsokolowskintl** [aristanetworks/EosSdk#29]
Hi,

I am trying to read current speed of an Ethernet interface and my agent either crashes or function link_speed throws an exception. If my agent is inheriting from eth_phy_intf_handler then I get SIGABRT and the application is aborting with following error in the logs:
_psmenet: eth_phy_intf.cpp:125: eos::eth_link_speed_t eos::convert(Interface::EthSpeed): Assertion `false' failed.
Unix Time: 1496930889: Process 2333 died with signal 6 (SIGABRT) errno 0 code -6 (unknown)
The killer was pid 2333 (myself) with uid 0
Backtrace: (most recent first)
/lib/libtac.so.0(ZN13SignalHandler13signalHandlerEiP9siginfo_tPv+0x21f)[0xeeb4d75f]_

If I don't inherit from the eth_phy_intf_handler and I am only using eos::sdk to get the port speed then an exception is thrown. Is there anything wrong with that code:
```
uint32_t EosAgent::get_speed(const string& intf_name) {
    try {
        eos::sdk_scoped_lock ssl;
        auto* mgr = sdk.get_eth_phy_intf_mgr();
        auto intf_id = sdk.get_intf_mgr()->eos_intf_name(intf_name);
        if (!sdk.get_intf_mgr()->exists(intf_id)) {
            log_error(GET_LOGGER("network-agent"),
                      "Eos Agent interface does not exist " + intf_name);
            return 0;
        }
        switch(mgr->link_speed(intf_id)) {
            case eos::LINK_SPEED_10MBPS:
                return 10;
            case eos::LINK_SPEED_100MBPS:
                return 100;
            case eos::LINK_SPEED_1GBPS:
                return 1000;
            case eos::LINK_SPEED_10GBPS:
                return 10000;
            case eos::LINK_SPEED_40GBPS:
                return 40000;
            case eos::LINK_SPEED_100GBPS:
                return 100000;
            case eos::LINK_SPEED_UNKNOWN:
            default:
                log_error(GET_LOGGER("network-agent"),
                          "Eos Agent returned unknown speed for " + intf_name);
                break;
        }
    }
    catch(eos::no_such_interface_error &error) {
        log_error(GET_LOGGER("network-agent"),
                  "Eos Agent no such interface " + intf_name);
    }
    catch(...) {
        log_error(GET_LOGGER("network-agent"),
                  "Eos Agent unknown exception when getting speed for "
                  + intf_name);
    }
    return 0;
}
```
The code inside catch(...) gets executed for each iterated Ethernet interface:
_2017-06-08 15:17:35.340225169 - ERROR - network-agent - Eos Agent unknown exception when getting speed for et16_1
2017-06-08 15:17:35.354120901 - ERROR - network-agent - Eos Agent unknown exception when getting speed for et33_

I am using EOS 4.18.

Regards,
Hubert

**Comment by ruferp** [comment]
Using the cli, what port speed does it show on your switch?
The assert is simply that the speed we read was not expected (we convert internal speed values to those exported in eos/types/eth_phy_intf.h). I agree that an assert is horse medicine in that case, returning LINK_SPEED_UNKNOWN would have been more appropriate...

**Comment by rmadsen** [comment]
Are you grabbing the reference to the mgr before starting EOS SDK's main loop?

**Comment by hubertsokolowskintl** [comment]
The CLI shows 10Gb/s for et33 so it seems correct:
arista0(config)#show interfaces et33
Ethernet33 is down, line protocol is notpresent (notconnect)
  Hardware is Ethernet, address is 2899.3a43.1edb (bia 2899.3a43.1edb)
  Ethernet MTU 9214 bytes , BW 10000000 kbit
  Full-duplex, 10Gb/s, auto negotiation: off, uni-link: n/a
  Loopback Mode : None
  0 link status changes since last clear
  Last clearing of "show interface" counters never
  5 minutes input rate 0 bps (0.0% with framing overhead), 0 packets/sec
  5 minutes output rate 0 bps (0.0% with framing overhead), 0 packets/sec
     0 packets input, 0 bytes
     Received 0 broadcasts, 0 multicast
     0 runts, 0 giants
     0 input errors, 0 CRC, 0 alignment, 0 symbol, 0 input discards
     0 PAUSE input
     0 packets output, 0 bytes
     Sent 0 broadcasts, 0 multicast
     0 output errors, 0 collisions
     0 late collision, 0 deferred, 0 output discards
     0 PAUSE output

What do you mean by getting reference? Is that line enough in the agent's constructor initialization list?
eos::eth_phy_intf_handler(sdk.get_eth_phy_intf_mgr())

**Comment by hubertsokolowskintl** [comment]
I just realized getting speed is causing agent abort only for interfaces /2 /3 and /4, i.e. et25/2. In CLI when I do show for such interfaces I get nothing:
arista0(config)#show interfaces et25/2
arista0(config)#

I tried using different methods to check if I can call link_speed but it does not work. I tried: exists from the manager, but it returns true for all 130 Ethernet interfaces. I tried hardware_present from eth_phy_intf manager but again it returns true for all interfaces. 
How do I check an interface before calling link_speed on it to make sure it will not abort my agent?

**Comment by hubertsokolowskintl** [comment]
I have changed EOS version on my Arista switch hoping to get better results but in fact it works even worse, now my application crashed with SIGSEGV when calling kernel_intf_name function from the manager class or when calling admin_enabled. my switch is Arista DCS-7060CX-32S-F, EOS version EOS-4.18.1FX-7060X.1-SSU.
I can avoid using kernel_intf_name but I need to be able to at least get administrative state of a port, so please help me with this issue. I have not tried reading the speed yet. My code for reading port admin state is this:
`bool EosAgent::is_admin_enabled(const string& intf_name) const {
    sdk_scoped_lock ssl;
    auto* mgr = get_intf_mgr();
    for(auto i = mgr->intf_iter(); i; ++i){
        if (i->to_string() == intf_name) {
            printf("calling admin enable\n");
            fflush(stdout);
            return mgr->admin_enabled(*i);
        }
    }
    printf("not found\n");
    fflush(stdout);
    return false;
`

Here are the logs:
calling admin enable
Unix Time: 1497369832: Process 24884 died with signal 11 (SIGSEGV) errno 0 code 1 (unknown)
The killer was pid 12865 with uid 0
 While accessing address 0x3241
Backtrace: (most recent first)
/lib/libtac.so.0(_ZN13SignalHandler13signalHandlerEiP9siginfo_tPv+0x21f)[0xeec1375f]
[0xf76e5490]
/lib/libeos.so.1(+0x19e938)[0xf7406938]

Regards,
Hubert

**Comment by ruferp** [comment]
I have tried calling admin_enabled() on EOS 4.18.2 using a DCS-7060CX-32S-F hardware and could not reproduce. I will try the exact same software version that you are using later. In the meantime: how are you compiling your application (gcc version)? Is the libeos version you use to compile against the same as the one on the switch? And, have you tried a python agent: does that fail too?

**Comment by ruferp** [comment]
I looked at the EosSdk version that comes with mint EOS-4.18.1FX-7060X.1-SSU, and that's 1.9.0. Did you use the 1.9.0 version of the EosSdk stubs to build your application? I'm confused, because on our software-download page I cannot see any EosSdk 1.9.0 version to download for EOS 4.18.1. If I build a test application with EosSdk 1.12.0, then it blows into pieces on admin_enabled() on a stock EOS-4.18.1FX-7060X.1-SSU, until I overwrite /usr/lib/libeos.so.1 with a 1.12.0 one. I also have no problem with a EosSdk 1.9.0 test application on EOS-4.18.1FX-7060X.1-SSU. If you use EosSdk 1.12.0 to build your application, can you also load the corresponding libeos.so on your switch?

**Comment by ruferp** [comment]
You could try those stubs that match the libeos.so in your image:
  ftp://ftp.arista.com/data/ar/eossdk-1.9.0-EOS-4.18.1/EosSdk-stubs-1.9.0.tar.gz
And you could overwritte /usr/lib/libeos.so.1.9.0 on the switch with the one below to fix the issue with link_speed:
  ftp://ftp.arista.com/data/ar/eossdk-1.9.0-EOS-4.18.1/libeos.so.1.9.0

**Comment by hubertsokolowskintl** [comment]
I was using SDK Stubs from the github because I though it is compatible with my version on the switch.
With this 1.9.0 stubs SDK admin_enabled works well. Also link_speed works after I uploaded the library from the second link to the switch. Are you planning to release an update to EOS-4.18.1FX-7060X image with that fix included?

Thanks for help!

Kind regards,
Hubert

**Comment by ruferp** [comment]
Thanks for confirming that the link_speed issue is fixed with the provided test libeos.so.
We will try to fix it in 4.18.3 (next release).
Not sure about the special SSU release you are using, maybe that SSU stuff will be rolled into stock 4.18.3, in which case it is a moot point. 
But that needs lobbying release managers, so outcome is not guarantied (I need to circle back here when that dust settles). 
For sure we can and should build 4.18.{1,2,3} versions of EosSdk.i686.rpm (which would be at EosSdk version 1.12.1) and push them onto our software download page where they are currently missing, our bad, sorry!

**Comment by hubertsokolowskintl** [comment]
I thought this SSU image is intended specifically for my switch. I also tried EOS-4.18.2F image from Active Releases folder but this OS does not list Ethernet interfaces even in CLI, only the management port is listed. I can try it again if you can help me with this problem. Also it would be better if you can provide the SDK in the same source tarball form as you did with 1.9.0 SDK.
Also I noticed that 1.9.0 SDK supports more port speed types then the SDK on github, i.e. it supports 25Gbps which is what my switch supports. Will the SDK for 4.18.{2,3} support it too?
Thanks,
Hubert

**Comment by ruferp** [comment]
I put the latest EosSdk code for the FX-7060-SSU release here:

ftp://ftp.arista.com/data/ar/eossdk-1.12.1-EOS-4.18.1-FX-7060X.2-SSU/EosSdk-stubs-1.12.1.tar.gz
ftp://ftp.arista.com/data/ar/eossdk-1.12.1-EOS-4.18.1-FX-7060X.2-SSU/EosSdk.i686.rpm

I used version .2 of the FX-7060-SSU release, I hope you don't mind (I think you are using .1).

This is EosSdk version 1.12.1. 
One notable difference in that version is that if you compile the stubbed libeos.so (by running the build.sh script in the stubs tarball), it will have a SONAME of libeos.so.<major>.<minor>, therefore any executable linked against it would not even start if the linker cannot find a corresponding (binary compatible) version on the switch (e.g. same library <major-version>.<minor-version>). The EosSdk.i686.rpm still has libeos.so.1 pointing to libeos.so.<major>.<minor>.<patch> for those people that don't link against the "correct" stubs...

Yes, the EosSdk-1.9.0 I gave you was supported 25Gbps, that's why it was not asserting... The version above has it too. 4.8.2 has shipped already (a month ago), and I think 4.8.3 is about to ship, so 4.8.4 would be the earliest, no date yet.

**Comment by hubertsokolowskintl** [comment]
Thanks for sharing the new SDK with me.
I don't see EOS-4.18.1-FX-7060X.2-SSU available on the software download page. There is only .1 version. Are you going to release it soon too?
Will 4.18.3 contain the fix for the link_speed issue and will it support 25 Gbps? I could switch to this image on my switch then.

Thanks,
Hubert

**Comment by ruferp** [comment]
Sorry, it looks like EOS-4.18.1-FX-7060X.2-SSU is missing from the download page, it will be added shortly. 
In the meantime I added an "enhanced"/convenient version of it, both EOS.swi and EOS-INT.swi (which one are you using?), into the same ftp directory as the previous EosSdk.i686.rpm. "Enhanced swi" in that it already includes the latest EosSdk.i686.rpm (with the link_speed fix amongst others). Use it for test only, since it is not an official version and we won't keep its debug-info (to debug crash dumps) along for long, etc...
The Tomahawk ASU2 special release you are using has been integrated into the upcoming 4.19 train (has not merged to 4.18.x, no known plans as far as I know).

**Comment by hubertsokolowskintl** [comment]
So I am assuming EOS-4.18.1-FX-7060X.2-SSU does not contain the fix for link_speed. Will 4.18.3 contain it? AFAIK the 4.18.X line supports my switch too and I don't need to use SSU releases actually.

**Comment by ruferp** [comment]
The official fx-2-ssu does not have the link-speed fix, only the one I put on ftp does, and that was an engineering build. 4.18.3 is about to ship and will not have the link-speed fix, that will come with 4.18.4.

**Comment by krzysztof-sprzaczkowski** [comment]
Are you able to provide us more detailed information when a new official version (4.18.4) of EOS is planned to be release? This is important for us to have such information this week. Thanks, Krzysztof

**Comment by ruferp** [comment]
The new version (4.18.4) will have the link_speed fix. We don't yet know when it will ship. Sometime after August 15.

**Comment by hubertsokolowskintl** [comment]
Hi,

I wanted to download again libeos with the fix from here but the FTP server is not responding.

ftp://ftp.arista.com/data/ar/eossdk-1.9.0-EOS-4.18.1/libeos.so.1.9.0

Is it going to be back soon?

Regards,
Hubert

**Comment by hubertsokolowskintl** [comment]
Thanks for putting it back.
Is 4.18.4 still planned to be released by the end of August?

regards,
Hubert

**Comment by krzysztof-sprzaczkowski** [comment]
Hi,
I'd like to back to date of a new EOS which is going to be released after August 15 (based on the latest info from ruferp). We have a strong dependency to your new EOS release in our project. So, we need to have more precise and official information about your release schedule in this particular case. Are you able to help to arrange direct contact with you (or someone from Arista's support) to talk about details? If so later I will be able to provide you contact info from our side.

Thanks,
Krzysztof

**Comment by hubertsokolowskintl** [comment]
Hi,
Please restore file
ftp://ftp.arista.com/data/ar/eossdk-1.9.0-EOS-4.18.1/EosSdk-stubs-1.9.0.tar.gz
I get ERROR 404: Not Found.

thanks,
Hubert

**Comment by ruferp** [comment]
Like said before, this was an engineering release, and I have deleted it, not just from the ftp server where it gets deleted automatically... (I would have to spend some time to redo it). 
However, EOS 4.18.4 is about to ship (probably tomorrow), and the corresponding EosSdk 1.12.3 is already up on https://www.arista.com/en/support/software-download (you need just the stubs tarball, the EosSdk.i686.rpm is already part of the EOS.swi that is about to be published).

**Comment by hubertsokolowskintl** [comment]
I have just migrated to this release and have one problem with the SDK. In version.m4 it is defined as 1.10.0:
m4_define([_NAME_],[EosSdk])
m4_define([_VERSION_],[1.10.0])
As a result libeos.so symbolic link points to libeos.so.1.10.0 after build and my application fails to start on the switch with error:
error while loading shared libraries: libeos.so.1.10: cannot open shared object file: No such file or directory
I think I can workaround this problem, just letting you know about it.

Thanks,
Hubert

**Comment by hubertsokolowskintl** [comment]
I confirm getting speed on 4.18.4 release works well. Thanks for fixing that issue. I think we can close this one.

**Comment by ruferp** [comment]
About the wrong version.m4: I think the github stubs tarball might have been uploaded too early. Try the one from https://www.arista.com/en/support/software-download (or just modify the m4_define to set 1.12.3). Besides the version, the other differences are mostly white-spaces and copyright blargh.
